### ISSUE TITLE
No response after change channel in TV Service

### ISSUE THREAD

**Issue by Kienz** [ebaauw/homebridge-zp#122]
Küche TV: set Active Identifier to "Alle Leut'"                                                                                                                                                                                
[1/11/2020, 11:07:44 PM] [ZP] error: Invalid character in entity name                                                                                                                                                                                        
Line: 0                                                                                                                                                                                                                                                      
Column: 643                                                                                                                                                                                                                                                  
Char:

**Comment by ebaauw** [comment]
I don't recognise that error message - it's not generated by homebridge-zp itself.  I would guess it's from the underlying library translating XML (from the zone players) to JSON (used by homebridge-zp).

I need to see the error in context.  Could you please capture and attach a full debug log file, see https://github.com/ebaauw/homebridge-hue#debug-log-file.

Does this happen for any channel, or only for `Alle Leut'`?  Is something choking on the single quote character `'` in the name?  What happens when you play that favourite from the Sonos app?  What source does that favourite come from?

**Comment by Kienz** [comment]
> [1/11/2020, 11:07:44 PM] [ZP] error: Invalid character in entity name
This comes from the  single quote character in the name.

But if I change channel to any channel I got "No response" after this action in HomeKit. In the logs it shows "[1/12/2020, 6:03:15 PM] [ZP] Büro TV: warning: set Active Identifier: timed out"

**Comment by ebaauw** [comment]
> This comes from the single quote character in the name.

The name of what?  I’ve got plenty of artist, album, and song names with a single quote character and I’ve never seen this error.

```
[1/12/2020, 6:03:15 PM] [ZP] Büro TV: warning: set Active Identifier: timed out
```
Homebridge-zp didn’t receive a timely response from your zone player after sending the command to change the input stream.  Could be an issue with your network or with the zone player.  Could also be that homebridge-zp is a bit too aggressive.  You might try and set a larger `timeout` in config.json.

**Comment by ebaauw** [comment]
Did some testing and I'm seeing the same warning as well.
```
Jan 14 19:57:15 pi2 homebridge[25033]: [Sonos] Home TV: set Active Identifier to "CITY FM Classic Rock"
Jan 14 19:57:15 pi2 homebridge[25033]: [Sonos] RINCON_000E58A8171201400 [Home]: request 881: POST /MediaRenderer/AVTransport/Control SetAVTransportURI
Jan 14 19:57:15 pi2 homebridge[25033]: [Sonos] RINCON_000E58A8171201400 [Home]: event: NOTIFY MediaRenderer/AVTransport/Event
Jan 14 19:57:15 pi2 homebridge[25033]: [Sonos] Home Sonos: ignore AVTransPortEvent after SetAVTransportURI
Jan 14 19:57:15 pi2 homebridge[25033]: [Sonos] RINCON_000E58A8171201400 [Home]: request 882: POST /MediaRenderer/AVTransport/Control Play
Jan 14 19:57:15 pi2 homebridge[25033]: [Sonos] RINCON_000E58A8171201400 [Home]: request 881: status 200
Jan 14 19:57:16 pi2 homebridge[25033]: [Sonos] Home TV: warning: set Active Identifier: timed out
Jan 14 19:57:16 pi2 homebridge[25033]: [Sonos] RINCON_000E58A8171201400 [Home]: request 882: status 200
Jan 14 19:57:16 pi2 homebridge[25033]: [Sonos] RINCON_000E58A8171201400 [Home]: event: NOTIFY MediaRenderer/AVTransport/Event
Jan 14 19:57:16 pi2 homebridge[25033]: [Sonos] Home Sonos: set Active Identifier from 6 to 7
Jan 14 19:57:16 pi2 homebridge[25033]: [Sonos] Living Room Sonos: set Active Identifier from 6 to 7
Jan 14 19:57:16 pi2 homebridge[25033]: [Sonos] Living Room TV: set Active Identifier to "CITY FM Classic Rock"
```

The error is actually generated by homebridge-lib.  When changing a characteristic from HomeKit, homebridge passes control to the plugin to update the device.  The plugin then responds to homebridge that the update has succeeded and homebridge informs HomeKit.  During this time, the HomeKit app is unresponsive.

As a safeguard against rogue plugins, homebridge-lib sets a timeout, hardcoded to 1000 ms (1s), for a plugin to handle the change.  When the plugin doesn't respond in that second, homebridge-lib issues the warning and returns the error to HomeKit, making the app responsive again.

When changing the input channel, homebridge-zp:
- Sends a `SetAVTransportURI` command to the zone player, to change it's source;
- Wait for confirmation from the zone player that it has accepted the command;
- Wait for the zone player to confirm that it has changed the source, through an `AVTransport` event;
- Sends a `Play` command to the zone player, to start playing the new input source.
- Wait for confirmation from the zone player that it has accepted the command;
- Returns control to homebridge (thru homebridge-lib).

This sequence might actually take longer than a second, when the new source is an Internet stream that needs to be buffered, before the zone player sends a response to the`Play` command.  Note that, in the example above, despite the warning and the error status reported to HomeKit, the input is changed alright, and the zone player actually plays the new input.

I've done some tests with the timeout set to 5000ms.  The above error no longer occurs, but there's another issue.  The code doesn't handle the case when the `AVTransportEvent` is handled before the response to the `SetAVTransportURI` command (which could happen due to asynchronous processing).  In this case, the `Play` command is never sent the timeout to occurs anyway, even after 5s.  This does explain my earlier observation, that sometimes the input switches, but doesn't play.

**Comment by ebaauw** [comment]
Above commit handles the timing of the `AVTransportEvent` and returns control to HomeKit after sending the `Play` command, instead of after receiving the response to the `Play` command.

I want to run this for some time, while deciding whether it's better to increase the timeout instead.  I haven't seen the timeout on issuing Play from HomeKit, so it really might be just this sequence of interactions.

**Comment by ebaauw** [comment]
It seems to be working OK; I just published the change in v1.0.20.

**Comment by Kienz** [comment]
Looks good!
